### PR TITLE
Electromagnetics PML example

### PR DESCRIPTION
**Pull request by psocratis**
The example demonstrates the use of a Cartesian Perfectly Matched Layer (PML) for simulating absorption for the indefinite Maxwell equations. Some examples are provided with exact solutions

a) 2D and 3D problems driven by a point source. The exact solution is given by a combination of Bessel functions. 
b) Lshape and Fichera problem. Same as (a) but taking advantage of symmetry.
c) Rectangular waveguide problem: Simulation of the first propagating mode

For the exact solutions of the PML examples see
https://link.springer.com/article/10.1007%2Fs00466-018-1640-3

- [x] Code builds.
- [x] Code passes `make style`.
- [x] Update `CHANGELOG`:
- [x] Update `.gitignore`:
- [ ] New examples:
    - [x] All sample runs at the top of the example work.
    - [x] Update `examples/makefile`:
      - [x] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [x] Add any files generated by it to the `clean` target.
      - [x] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [x] Update `examples/CMakeLists.txt`:
      - [x] Add the example code to the `ALL_EXE_SRCS` variable.
      - [x] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [x] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1279](https://github.com/mfem/mfem/pull/1279) | @psocratis | @tzanio | @mlstowell + @dylan-copeland | 02/16/20 | 03/16/20 | 03/25/20 | |

### CODE REVIEW

**Comment by artv3** [comment]
Mind if I try this out when its ready? I like these kinds of problems 👍

**Comment by psocratis** [comment]
> Mind if I try this out when it is ready? I like these kinds of problems

Not at all. It still needs some final adjustments before it's ready. I'll be happy to hear any suggestions you might have

File: examples/ex23.cpp [hunk 1]
```diff
@@ -0,0 +1,988 @@
...
+//               The example also demonstrates the use of complex valued bilinear and linear forms.
+//               We recommend viewing example 22 before viewing this example.
+//               Four examples are probided with exact solutions (iprob = 0-4).
+
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+#ifdef _WIN32
+#define jn(n, x) _jn(n, x)
```
Comment (dylan-copeland) [inline examples/ex23.cpp]: Also yn.

File: examples/ex23.cpp [hunk 2]
```diff
@@ -0,0 +1,988 @@
...
+   args.PrintOptions(cout);
+
+   Mesh * mesh = new Mesh(mesh_file, 1, 1);
+   dim = mesh->Dimension();
+
+   //Angular frequency
+   omega = 2.0 * M_PI * freq;
+
+   // Setup PML length
+   Array2D<double> lngth(dim, 2);  lngth = 0.0;
```
Comment (dylan-copeland) [inline examples/ex23.cpp]: Why not name this variable "length" for better readability?

File: examples/ex23.cpp [hunk 3]
```diff
@@ -0,0 +1,988 @@
...
+      sol_i_ofs.precision(8);
+      x.real().Save(sol_r_ofs);
+      x.imag().Save(sol_i_ofs);
+   }
+
+   // 17. Send the solution by socket to a GLVis server.
+   if (visualization)
+   {
+      string keys;
+      keys = (dim == 3) ? "keys macF\n" : keys = "keys amrRljcUUuu\n";
```
Comment (dylan-copeland) [inline examples/ex23.cpp]: Why the random strings?

File: examples/ex23.cpp [hunk 4]
```diff
@@ -0,0 +1,988 @@
...
+
+         if (dim == 2)
+         {
+            double x0 = x(0) + shift(0);
+            double x1 = x(1) + shift(1);
+            double r = sqrt(x0 * x0 + x1 * x1);
+            double beta = k * r;
+
+            // Bessel functions
+            complex<double> Ho = jn(0, beta) + zi * yn(0, beta);
```
Comment (dylan-copeland) [inline examples/ex23.cpp]: There is inconsistent use of `std::complex<double>` and `complex<double>`. You could just use the latter everywhere.

File: examples/ex23.cpp [hunk 5]
```diff
@@ -0,0 +1,988 @@
...
+
+   for (int i = 0; i < dim; ++i)
+   {
+      det *= dxs[i];
+   }
+
+   M = 0.0;
+   for (int i = 0; i < dim; ++i)
+   {
+      M(i, i) = (det / pow(dxs[i], 2)).imag();
```
Comment (dylan-copeland) [inline examples/ex23.cpp]: It may or may not make much difference, depending on where it is used, but I generally avoid pow() when possible, i.e. when the power is a small integer. In a bottleneck, pow() can slow down a code significantly.

File: examples/ex23.cpp [hunk 6]
```diff
@@ -0,0 +1,988 @@
...
+
+   for (int i = 0; i < dim; ++i)
+   {
+      det *= dxs[i];
+   }
+
+   M = 0.0;
+   for (int i = 0; i < dim; ++i)
+   {
+      M(i, i) = abs(det / pow(dxs[i], 2));
```
Comment (dylan-copeland) [inline examples/ex23.cpp]: There is some ambiguity in the type used by abs(). Although this works, it may be clearer to use fabs or std::abs.

**Comment by dylan-copeland** [comment]
The test `ex23 -o 2 -f 1.0 -ref 3 -prob 3` is quite slow. It could be replaced with `ex23 -o 2 -f 1.0 -ref 2 -prob 3` which is must faster and gives similar results.

Also, some lines of code are long.

File: examples/ex23.cpp [hunk 7]
```diff
@@ -0,0 +1,988 @@
...
+double mu = 1.0;
+double epsilon = 1.0;
+double omega;
+int dim;
+bool exact_known = false;
+
+enum prob_type
+{
+   beam,     // PML on one end of the domain
+   scatter,  // Scattering in a square or a cube
```
Comment (dylan-copeland) [inline examples/ex23.cpp]: "Scattering from"?

File: examples/ex23p.cpp [hunk 8]
```diff
@@ -0,0 +1,1047 @@
+//                       MFEM Example 23 - Parallel Version
+//
+// Compile with: make ex23p
+//
+// Sample runs:  mpirun -np 4 ex23p -o 2 -f 1.0 -rs 1 -rp 1 -prob 0
+//               mpirun -np 4 ex23p -o 3 -f 1.0 -rs 1 -rp 1 -prob 1
+//               mpirun -np 4 ex23p -o 2 -f 3.0 -rs 3 -rp 1 -prob 2
+//               mpirun -np 4 ex23p -o 2 -f 1.0 -rs 2 -rp 1 -prob 3
```
Comment (dylan-copeland) [inline examples/ex23p.cpp]: This is slow. Reducing to `-rs 1` is faster and still accurate enough.

File: examples/ex23.cpp (hunk 2 above)
Comment (psocratis) [reply examples/ex23.cpp]: renamed

File: examples/ex23.cpp (hunk 3 above)
Comment (psocratis) [reply examples/ex23.cpp]: The keys are not really random. They are chosen depending on the problem in order to show the most interesting characteristics of the solution

File: examples/ex23.cpp (hunk 4 above)
Comment (psocratis) [reply examples/ex23.cpp]: fixed

File: examples/ex23.cpp (hunk 6 above)
Comment (psocratis) [reply examples/ex23.cpp]: fabs will not work here since the argument is complex<double>

File: examples/ex23.cpp (hunk 7 above)
Comment (psocratis) [reply examples/ex23.cpp]: Fixed

File: examples/ex23p.cpp (hunk 8 above)
Comment (psocratis) [reply examples/ex23p.cpp]: fixed

File: examples/ex23.cpp (hunk 1 above)
Comment (psocratis) [reply examples/ex23.cpp]: Good catch! I added it!

File: examples/ex23.cpp (hunk 3 above)
Comment (dylan-copeland) [reply examples/ex23.cpp]: Is this the first example using keys like this in the visualization? If so, it would be informative to explain the usage in comments.

File: examples/ex25.cpp [hunk 9]
```diff
@@ -0,0 +1,989 @@
...
+   MatrixRestrictedCoefficient restr_c2_Re(c2_Re,attrPML);
+   MatrixRestrictedCoefficient restr_c2_Im(c2_Im,attrPML);
+
+   a.AddDomainIntegrator(new CurlCurlIntegrator(restr_c1_Re),
+                         new CurlCurlIntegrator(restr_c1_Im));
+   a.AddDomainIntegrator(new VectorFEMassIntegrator(restr_c2_Re),
+                         new VectorFEMassIntegrator(restr_c2_Im));
+
+   // 12. Assemble the parallel bilinear form and the corresponding linear
+   //     system, applying any necessary transformations such as: parallel
```
Comment (dylan-copeland) [inline examples/ex25.cpp]: Remove "parallel assembly."

File: examples/ex25.cpp [hunk 10]
```diff
@@ -0,0 +1,989 @@
...
+   //     assembly, eliminating boundary conditions, applying conforming
+   //     constraints for non-conforming AMR, etc.
+   a.Assemble();
+
+   OperatorHandle Ah;
+   Vector B, X;
+   a.FormLinearSystem(ess_tdof_list, x, b, Ah, X, B);
+
+   // 13. Transform to monolithic HypreParMatrix
```
Comment (dylan-copeland) [inline examples/ex25.cpp]: SparseMatrix

File: examples/ex25.cpp [hunk 11]
```diff
@@ -0,0 +1,989 @@
...
+
+      PmlMatrixCoefficient pml_c1_abs(cdim,detJ_inv_JT_J_abs, pml);
+      ScalarMatrixProductCoefficient c1_abs(muinv,pml_c1_abs);
+      MatrixRestrictedCoefficient restr_c1_abs(c1_abs,attrPML);
+
+      PmlMatrixCoefficient pml_c2_abs(dim, detJ_JT_J_inv_abs,pml);
+      ScalarMatrixProductCoefficient c2_abs(absomeg,pml_c2_abs);
+      MatrixRestrictedCoefficient restr_c2_abs(c2_abs,attrPML);
+
+      // Sesquilinear form inside the PML region
```
Comment (dylan-copeland) [inline examples/ex25.cpp]: This comment could be used for `a`.

**Comment by tzanio** [comment]
@mlstowell, do you need more time to review this?

File: examples/ex25.cpp (hunk 11 above)
Comment (psocratis) [reply examples/ex25.cpp]: Fixed comment!

File: examples/ex25.cpp (hunk 10 above)
Comment (psocratis) [reply examples/ex25.cpp]: Fixed

File: examples/ex25.cpp (hunk 9 above)
Comment (psocratis) [reply examples/ex25.cpp]: Removed

File: examples/ex25.cpp [hunk 12]
```diff
@@ -0,0 +1,991 @@
...
+
+// Description:  This example code solves a simple electromagnetic wave
+//               propagation problem corresponding to the second order indefinite
+//               Maxwell equation (1/mu) * curl curl E - \omega^2 * epsilon E = f
+//               with a Perfectly Matched Layer (PML).
+//               We discretize with Nedelec finite elements in 2D or 3D.
+//
+//               The example also demonstrates the use of complex valued bilinear and linear forms.
+//               We recommend viewing example 22 before viewing this example.
+//               Four examples are probided with exact solutions (iprob = 0-4).
```
Comment (mlstowell) [inline examples/ex25.cpp]: Doesn't 0-4 lead to 5 examples?

File: examples/ex25p.cpp [hunk 13]
```diff
@@ -0,0 +1,1050 @@
...
+
+// Description:  This example code solves a simple electromagnetic wave
+//               propagation problem corresponding to the second order indefinite
+//               Maxwell equation (1/mu) * curl curl E - \omega^2 * epsilon E = f
+//               with a Perfectly Matched Layer (PML).
+//               We discretize with Nedelec finite elements in 2D or 3D.
+//
+//               The example also demonstrates the use of complex valued bilinear and linear forms.
+//               We recommend viewing example 22 before viewing this example.
+//               Four examples are provided with exact solutions (iprob = 0-3).
```
Comment (mlstowell) [inline examples/ex25p.cpp]: It looks like `prob` = 0-4

File: examples/ex25p.cpp [hunk 14]
```diff
@@ -0,0 +1,1050 @@
...
+
+void E_bdr_data_Re(const Vector &x, Vector &E);
+void E_bdr_data_Im(const Vector &x, Vector &E);
+
+void E_exact_Re(const Vector &x, Vector &E);
+void E_exact_Im(const Vector &x, Vector &E);
+
+void source(const Vector &x, Vector & f);
+
+void detJ_JT_J_inv_Re(const Vector &x, CartesianPML * pml, DenseMatrix &M);
```
Comment (mlstowell) [inline examples/ex25p.cpp]: It's not clear what `J` refers to.  I think I know what you mean but some explanation would be useful.  Perhaps the better place to document this is above the `PmlMatrixCoefficient`.  We could get away with a brief explanation of the concept along with a reference to a standard publication.

File: examples/ex25.cpp [hunk 15]
```diff
@@ -0,0 +1,991 @@
+//                       MFEM Example 25 - Serial Version
```
Comment (mlstowell) [inline examples/ex25.cpp]: Try to reduce code width to 80 characters.  Lines in the "Sample runs" section can be longer than 80 characters but other comments and code should under this limit whenever possible.

File: examples/ex25p.cpp [hunk 16]
```diff
@@ -0,0 +1,1050 @@
+//                       MFEM Example 25 - Parallel Version
```
Comment (mlstowell) [inline examples/ex25p.cpp]: Again, try to reduce code width to 80 characters.  Lines in the "Sample runs" section can be longer than 80 characters but other comments and code should under this limit whenever possible.

**Comment by dylan-copeland** [comment]
I think we should put a reference to a publication explaining the PML choices (stretch function, etc.) in the code as a comment. The reference given in the description of this PR is fine.

File: doc/CodeDocumentation.dox [hunk 17]
```diff
@@ -84,6 +84,8 @@ namespace mfem {
  * - <a class="el" href="ex22_8cpp_source.html">Example 22</a>: complex-valued linear systems for damped harmonic oscillators
  * - <a class="el" href="ex22p_8cpp_source.html">Example 22p</a>: parallel complex-valued linear systems for damped harmonic oscillators
  * - <a class="el" href="ex23_8cpp_source.html">Example 23</a>: 2nd order in time the wave equation
+ * - <a class="el" href="ex25_8cpp_source.html">Example 24</a>: simulation of electromagnetic wave propagation using a Perfectly Matched Layer (PML)
+ * - <a class="el" href="ex25p_8cpp_source.html">Example 24p</a>: parallel simulation of electromagnetic wave propagation using a Perfectly Matched Layer (PML)
```
[the remark this answers is not in the document]
Comment (psocratis) [reply doc/CodeDocumentation.dox]: Fixed doc

File: examples/ex25.cpp (hunk 12 above)
Comment (psocratis) [reply examples/ex25.cpp]: Yes! Comment changed

File: examples/ex25p.cpp (hunk 13 above)
Comment (psocratis) [reply examples/ex25p.cpp]: Fixed

File: examples/ex25p.cpp (hunk 14 above)
Comment (psocratis) [reply examples/ex25p.cpp]: Added brief explanation

File: examples/ex25.cpp (hunk 15 above)
Comment (psocratis) [reply examples/ex25.cpp]: Checked and fixed all the line lengths

File: examples/ex25p.cpp (hunk 16 above)
Comment (psocratis) [reply examples/ex25p.cpp]: Checked and fixed all the line lengths

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
LGTM

File: examples/ex22.cpp [hunk 18]
```diff
@@ -382,6 +382,7 @@ int main(int argc, char *argv[])
       {
          case 0:
             pc_r = new DSmoother(*PCOp.As<SparseMatrix>());
+            s = 1.0;
```
Comment (tzanio) [inline examples/ex22.cpp]: Why this change?
